src/migratus/cli.clj

```diff
@@ -95,17 +98,109 @@
 
       :else (no-match-message args summary))))
 
+(defn util-date-to-local-datetime [util-date]
+  (if (= util-date nil)
+    nil
+    (let [instant (.toInstant util-date)
+          zone-id (ZoneId/systemDefault)
+          local-datetime (.atZone instant zone-id)]
+      local-datetime)))
+
+(defn formatted-date [date]
+  (if (nil? date)
+    nil
+    (core/format "%1tFT%<tTZ", date)))
+
+(defn parsed-migrations-data [cfg]
+  (let [all-migrations (migratus/all-migrations cfg)]
+    (map (fn [m] (let [{:keys [id name applied]} m
+                       to-local-date (if (nil? applied)
+                                       nil (util-date-to-local-datetime applied))
+                       date-str (formatted-date to-local-date)]
+                   {:id id :name name :applied date-str})) all-migrations)))
+
+(defn pending-migrations [cfg]
+  (filter (fn [mig] (= nil (:applied mig))) (parsed-migrations-data cfg)))
+
+(defn applied-migrations [cfg]
+  (filter (fn [mig] (not= nil (:applied mig))) (parsed-migrations-data cfg)))
+
+(defn cli-print-migrations-edn! [data]
+  (log/info data))
+
+(defn cli-print-migrations-json! [data]
+  (log/info (cheshire/generate-string data)))
+
+(defn cli-print-migrations! [data f]
```

Comment:
I think you can redo the whole printing of json and edn. 
You dance around the issue a lot with not much benefit. 

For output: 
You wither build the whole data once as a string and dump it, or you convert each row and print that. 

You can iterate through rows, you do that for plain output. 
You can do the same for edn / json . 

You can also use lazy evaluation in this case to make a list that renders it's row in the format of choosing.